src/gluonnlp/metric/masked_accuracy.py

```diff
@@ -0,0 +1,304 @@
+        else:
+            pred = list(pred.values())
+
+        if self.label_names is not None:
+            label = [label[name] for name in self.label_names]
+        else:
+            label = list(label.values())
+
+        self.update(label, pred)
+
+    def update(self, labels, preds):
+        """Updates the internal evaluation result.
+
+        Parameters
+        ----------
+        labels : list of `NDArray`
+            The labels of the data.
+
+        preds : list of `NDArray`
+            Predicted values.
+        """
+        raise NotImplementedError()
+
+    def reset(self):
+        """Resets the internal evaluation result to initial state."""
+        self.num_inst = 0
+        self.sum_metric = 0.0
+        self.global_num_inst = 0
+        self.global_sum_metric = 0.0
+
+    def reset_local(self):
+        """Resets the local portion of the internal evaluation results
+        to initial state."""
+        self.num_inst = 0
+        self.sum_metric = 0.0
+
+    def get(self):
+        """Gets the current evaluation result.
+
+        Returns
+        -------
+        names : list of str
+           Name of the metrics.
+        values : list of float
+           Value of the evaluations.
+        """
+        if self.num_inst == 0:
+            return (self.name, float('nan'))
+        else:
+            return (self.name, self.sum_metric / self.num_inst)
+
+    def get_global(self):
+        """Gets the current global evaluation result.
+
+        Returns
+        -------
+        names : list of str
+           Name of the metrics.
+        values : list of float
+           Value of the evaluations.
+        """
+        if self._has_global_stats:
+            if self.global_num_inst == 0:
+                return (self.name, float('nan'))
+            else:
+                return (self.name, self.global_sum_metric / self.global_num_inst)
+        else:
+            return self.get()
+
+    def get_name_value(self):
+        """Returns zipped name and value pairs.
+
+        Returns
+        -------
+        list of tuples
+            A (name, value) tuple list.
+        """
+        name, value = self.get()
+        if not isinstance(name, list):
+            name = [name]
+        if not isinstance(value, list):
+            value = [value]
+        return list(zip(name, value))
+
+    def get_global_name_value(self):
+        """Returns zipped name and value pairs for global results.
+
+        Returns
+        -------
+        list of tuples
+            A (name, value) tuple list.
+        """
+        if self._has_global_stats:
+            name, value = self.get_global()
+            if not isinstance(name, list):
+                name = [name]
+            if not isinstance(value, list):
+                value = [value]
+            return list(zip(name, value))
+        else:
+            return self.get_name_value()
+
+
+class MaskedAccuracy(EvalMetric):
```

Comment:
This is useful, we can implement this by inheriting from `mxnet.metric.EvalMetric`.

Reply:
I am not sure for now. Since gluonnlp is built on night build of mxnet.
We can always 
`pip install --pre --upgrade mxnet`
then get the newly version which contains `mxnet.metric`.

Then in gluonnlp, we can simply use:
```
from mxnet.metric import check_label_shapes
from mxnet.metric import EvalMetric
```

If there is a plan in mxnet release to remove this api, could you specify a little bit more details which are the upstream interface in mxnet might be impacted and why there is such change?

@szha could you also please take a look at this?

Reply:
GluonNLP currently depends on the latest release instead of nightly builds.

Reply:
To elaborate more, the pre-training PR uses the `reset_local` API of EvalMetric which is only available on mxnet master:
```
➜  Documents python
Python 2.7.13 (default, Dec 18 2016, 07:03:39)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import mxnet as mx
>>> mx.__version__
'1.3.0'
>>> mx.metric.EvalMetric
<class 'mxnet.metric.EvalMetric'>
>>> mx.metric.EvalMetric.reset
<unbound method EvalMetric.reset>
>>> mx.metric.EvalMetric.reset_local
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'EvalMetric' has no attribute 'reset_local'
>>>
```

Reply:
Thanks. What's the reason of using `reset_local`  API, not the `reset` one?

Reply:
And this is different from the previous discussion on the mxnet release doesn't have the API. That's the reason I want to clarify and makes this confused.

Reply:
And what's the reason we only change to use `mx.metric.check_label_shapes` not `mx.metric. EvalMetric `. Since `mx.metric.check_label_shapes` has neither not `reset_local ` nor `reset` attribute. I am very confused now...

```
>>> mx.__version__
'1.3.0'
>>> mx.metric.EvalMetric
<class 'mxnet.metric.EvalMetric'>
>>> mx.metric.EvalMetric.reset
<function EvalMetric.reset at 0x108e882f0>
>>> mx.metric.EvalMetric.reset_local
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'EvalMetric' has no attribute 'reset_local'
>>> mx.metric.check_label_shapes
<function check_label_shapes at 0x108e5dea0>
>>> mx.metric.check_label_shapes.reset
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'function' object has no attribute 'reset'
>>> mx.metric.check_label_shapes.reset_local
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'function' object has no attribute 'reset_local'
```

Reply:
`mx.metric.EvalMetric` doesn't have `reset_local` method, and that is used in gluonnlp. 
It resets the local portion of the metric. For example, say I want to print the accuracy per 10 batches, as well as the accuracy for the full epoch. 

Normally one would create 2 metrics m1 and m2, call m1.update and m2.update per batch, and call m1.reset per 10 batches, call m2.reset per epoch. 

Instead, now I can create one metric and call m.update per batch, m.reset_local() every 10 batches, and m.reset per epoch.